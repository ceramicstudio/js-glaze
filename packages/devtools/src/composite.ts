import type { CeramicApi } from '@ceramicnetwork/common'
import type {
  CompositeViewsDefinition,
  EncodedCompositeDefinition,
  InternalCompositeDefinition,
  Model,
  ModelDefinition,
  RuntimeCompositeDefinition,
  StreamCommits,
} from '@glazed/types'
import type { GraphQLSchema } from 'graphql'
import cloneDeep from 'lodash-es/cloneDeep.js'
import merge from 'lodash-es/merge.js'
import createObjectHash from 'object-hash'

import { decodeSignedMap, encodeSignedMap } from './formats/json.js'
import { createRuntimeDefinition } from './formats/runtime.js'
import { compositeModelsAndCommonEmbedsFromGraphQLSchema } from './schema.js'

type StrictCompositeDefinition = Required<InternalCompositeDefinition>

function toStrictDefinition(definition: InternalCompositeDefinition): StrictCompositeDefinition {
  return {
    aliases: {},
    commonEmbeds: [],
    views: { account: {}, root: {}, models: {} },
    ...definition,
  }
}

function isSupportedVersion(supported: string, check: string): boolean {
  const [supportedMajor] = supported.split('.')
  const [checkMajor] = check.split('.')
  return supportedMajor === checkMajor
}

function assertSupportedVersion(supported: string, check: string): void {
  if (!isSupportedVersion(supported, check)) {
    throw new Error(`Unsupported Composite version ${check}, expected version ${supported}`)
  }
}

function assertModelsHaveCommits(
  models: Record<string, ModelDefinition>,
  commits: Record<string, StreamCommits>
): void {
  for (const id of Object.keys(models)) {
    if (commits[id] == null) {
      throw new Error(`Missing commits for model ${id}`)
    }
  }
}

export function setDefinitionAliases(
  definition: StrictCompositeDefinition,
  aliases: Record<string, string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.aliases
  definition.aliases = { ...existing, ...aliases }
  return definition
}

export function setDefinitionCommonEmbeds(
  definition: StrictCompositeDefinition,
  names: Iterable<string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? [] : definition.commonEmbeds
  definition.commonEmbeds = Array.from(new Set([...existing, ...names]))
  return definition
}

export function setDefinitionViews(
  definition: StrictCompositeDefinition,
  views: CompositeViewsDefinition,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.views
  definition.views = merge(existing, views)
  return definition
}

async function loadModelsFromCommits<Models = Record<string, StreamCommits>>(
  _ceramic: CeramicApi,
  modelsCommits: Models
): Promise<Record<keyof Models, ModelDefinition>> {
  const definitions = await Promise.all(
    Object.values(modelsCommits).map((_commits): Promise<Model> => {
      throw new Error('Not implemented')
      // return await Model.fromCommits(ceramic, commits)
    })
  )
  return Object.keys(modelsCommits).reduce((acc, id, index) => {
    const model = definitions[index]
    if (model == null) {
      throw new Error(`Missing Model for id ${id}`)
    }
    const modelID = model.id.toString()
    if (modelID !== id) {
      throw new Error(`Unexpected Model stream ID: expected ${id}, got ${modelID}`)
    }
    acc[modelID as keyof Models] = model.content
    return acc
  }, {} as Record<keyof Models, ModelDefinition>)
}

export type CompositeParams = {
  commits: Record<string, StreamCommits>
  definition: InternalCompositeDefinition
}

export type CompositeInput = Composite | CompositeParams
export type CompositeOptions = {
  aliases?: Record<string, string>
  commonEmbeds?: 'all' | 'none' | Iterable<string>
  views?: CompositeViewsDefinition
}

export type CreateParams = {
  ceramic: CeramicApi
  schema: string | GraphQLSchema
}

export type FromJSONParams = {
  ceramic: CeramicApi
  definition: EncodedCompositeDefinition
}

export type FromModelsParams = CompositeOptions & {
  ceramic: CeramicApi
  models: Array<string>
}

export class Composite {
  static VERSION = '1.0'

  static create(params: CreateParams): Promise<Composite> {
    // TODO: convert the schema to models, create the models on the Ceramice node, load their commits

    const { models, commonEmbeds } = compositeModelsAndCommonEmbedsFromGraphQLSchema(params.schema)
    console.log('MODELS', models)
    console.log('COMMON EMBEDS', commonEmbeds)

    throw new Error('Not implemented')
  }

  static from(composites: Iterable<CompositeInput>, options?: CompositeOptions): Composite {
    const [first, ...rest] = composites
    if (first == null) {
      throw new Error('Missing composites to compose')
    }
    const composite = first instanceof Composite ? first.clone() : new Composite(first)
    return composite.merge(rest, options)
  }

  static async fromJSON(params: FromJSONParams): Promise<Composite> {
    const { models, ...definition } = params.definition
    const commits = decodeSignedMap(models)
    return new Composite({
      commits,
      definition: {
        ...definition,
        models: await loadModelsFromCommits(params.ceramic, commits),
      },
    })
  }

  static fromModels(_params: FromModelsParams): Promise<Composite> {
    // TODO: load all models definitions and commits
    throw new Error('Not implemented')
  }

  #commits: Record<string, StreamCommits>
  #definition: StrictCompositeDefinition
  #hash: string | undefined

  constructor(params: CompositeParams) {
    assertSupportedVersion(Composite.VERSION, params.definition.version)
    assertModelsHaveCommits(params.definition.models, params.commits)
    this.#commits = cloneDeep(params.commits)
    this.#definition = toStrictDefinition(cloneDeep(params.definition))
  }

  get hash(): string {
    if (this.#hash == null) {
      this.#hash = createObjectHash(this.#definition)
    }
    return this.#hash
  }

  clone(): Composite {
    return new Composite(this.toParams())
  }

  copy(models: Array<string>): Composite {
    if (models.length === 0) {
      throw new Error('Missing models to copy')
    }

    const { commits, definition } = this.toParams()
    const def = toStrictDefinition(definition)

    const nameIDs = Object.entries(def.models).reduce((acc, [id, model]) => {
      acc[model.name] = id
      return acc
    }, {} as Record<string, string>)
    const aliasIDs = Object.entries(def.aliases).reduce((acc, [id, alias]) => {
      acc[alias] = id
      return acc
    }, {} as Record<string, string>)

    const nextCommits: Record<string, StreamCommits> = {}
    const nextModels: Record<string, ModelDefinition> = {}
    const nextAliases: Record<string, string> = {}
    const nextViews: CompositeViewsDefinition = { account: {}, root: {}, models: {} }

    for (const model of models) {
      const id = aliasIDs[model] ?? nameIDs[model] ?? model
      if (def.models[id] == null) {
        throw new Error(`Model not found: ${model}`)
      }

      nextCommits[id] = commits[id]
      nextModels[id] = def.models[id]
      if (def.aliases[id] != null) {
        nextAliases[id] = def.aliases[id]
      }
      if (def.views.models[id] != null) {
        // TODO: check relations to other models, ensure there's no missing model in the subset
        nextViews.models[id] = def.views.models[id]
      }
      // TODO: account and root views
    }

    return new Composite({
      commits: nextCommits,
      definition: {
        version: def.version,
        commonEmbeds: def.commonEmbeds,
        models: nextModels,
        aliases: nextAliases,
        views: nextViews,
      },
    })
  }

  equals(other: CompositeInput): boolean {
    const otherHash =
      other instanceof Composite
        ? other.hash
        : createObjectHash(toStrictDefinition(other.definition))
    return this.hash === otherHash
  }

  merge(other: CompositeInput | Array<CompositeInput>, options: CompositeOptions = {}): Composite {
    const nextParams = this.toParams()
    const nextDefinition = toStrictDefinition(nextParams.definition)
    const collectedEmbeds = new Set<string>()

    for (const composite of Array.isArray(other) ? other : [other]) {
      const { commits, definition } =
        composite instanceof Composite ? composite.toParams() : composite
      assertSupportedVersion(nextDefinition.version, definition.version)
      assertModelsHaveCommits(definition.models, commits)

      const def = toStrictDefinition(definition)
      Object.assign(nextParams.commits, commits)
      Object.assign(nextDefinition.models, definition.models)
      Object.assign(nextDefinition.aliases, def.aliases)
      merge(nextDefinition.views, def.views)
      for (const name of def.commonEmbeds) {
        collectedEmbeds.add(name)
      }
    }

    if (options.aliases != null) {
      setDefinitionAliases(nextDefinition, options.aliases)
    }
    const commonEmbeds = options.commonEmbeds ?? 'none'
    if (commonEmbeds === 'all') {
      setDefinitionCommonEmbeds(nextDefinition, collectedEmbeds)
    } else if (Array.isArray(commonEmbeds)) {
      setDefinitionCommonEmbeds(nextDefinition, commonEmbeds, true)
    }
    if (options.views != null) {
      setDefinitionViews(nextDefinition, options.views)
    }

    return new Composite({ ...nextParams, definition: nextDefinition })
  }

  setAliases(aliases: Record<string, string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionAliases(toStrictDefinition(params.definition), aliases, replace)
    return new Composite({ ...params, definition })
  }

  setCommonEmbeds(names: Iterable<string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionCommonEmbeds(
      toStrictDefinition(params.definition),
      names,
      replace
    )
    return new Composite({ ...params, definition })
  }

  setViews(views: CompositeViewsDefinition, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionViews(toStrictDefinition(params.definition), views, replace)
    return new Composite({ ...params, definition })
  }

  toJSON(): EncodedCompositeDefinition {
    return {
      version: this.#definition.version,
      models: encodeSignedMap(this.#commits),
      aliases: this.#definition.aliases,
      views: this.#definition.views,
    }
  }

  toParams(): CompositeParams {
    return {
      commits: cloneDeep(this.#commits),
      definition: cloneDeep(this.#definition),
    }
  }

  toRuntime(): RuntimeCompositeDefinition {
    return createRuntimeDefinition(this.#definition)
  }
}
