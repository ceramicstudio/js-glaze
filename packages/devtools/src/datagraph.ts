export class DataGraph {
  // Deterministic account to single record (IDX)
  async getRecord(type: string, did?: string): Promise<any> {}
  async setRecord(type: string, content: any): Promise<any> {}
  async removeRecord(type: string): Promise<any> {}

  // Account to multiple records
  async listCollectionRecords(type: string, did?: string): Promise<any> {}
  async addCollectionRecord(type: string, content: any): Promise<any> {}
  async removeCollectionRecord(id: string): Promise<any> {}

  // Deterministic account links
  async getAccountLink(type: string, did: string): Promise<any> {}
  async setAccountLink(type: string, did: string, content?: any): Promise<any> {}
  async removeAccountLink(type: string, did: string): Promise<any> {}

  // Account links query
  async listAccountLinksFrom(type: string, did?: string): Promise<any> {}
  async listAccountLinksTo(type: string, did?: string): Promise<any> {}
  async listReciprocalAccountLinks(type: string, did?: string): Promise<any> {}
  async listAccountLinksRecords(
    linkType: string,
    collectionType: string,
    did?: string
  ): Promise<any> {}

  // Records associated to stream -- should these use the CollectionRecords APIs?
  // Need to define query parameters for collections
  async listChildrenRecords(streamID: string, type: string): Promise<any> {}
  async addChildRecord(streamID: string, type: string, content: any): Promise<any> {}
  // Should this simply be removeCollectionRecord?
  async removeChildRecord(streamID: string, type: string, id: string): Promise<any> {}

  // TODO: getCollection with all basic profiles
}
