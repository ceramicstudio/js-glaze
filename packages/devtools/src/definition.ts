export type DefinitionNodeType =
  | 'self' // default
  | 'did'
  | 'streamid'
  | 'any'

export type DefinitionNodeConfig = {
  type: DefinitionNodeType
  optional?: boolean // defaults to `false`
}

export type DefinitionConfig = {
  multiple?: boolean // defaults to `false` = deterministic stream (IDX default)
  to?: DefinitionNodeConfig
  from?: DefinitionNodeConfig
}

function isSelfNodeConfig(config?: DefinitionNodeConfig): boolean {
  return config == null || config.type === 'self'
}

// definition queries/mutations based on config type:
// - if multiple == false or not set, and to and from are "self" or unset -> IDX deterministic get/set
// - if multiple == false or not set

export function getDefinitionAccessPattern(config: DefinitionConfig = {}) {
  const isFromSelf = isSelfNodeConfig(config.from)
  const isToSelf = isSelfNodeConfig(config.to)

  if (config.multiple) {
    if (isFromSelf) {
      if (isToSelf) {
        // Collection list/add/remove with content on stream, ex: posts
      } else {
        // Collection list/add/remove linkTo, edge content based on schema, ex: comments on post
        // Could be optional link - in that case similar to previous logic
      }
    } else if (isToSelf) {
      // Collection of links with specified subject but object is stream content - what use case?
    } else {
      // Collection of links with specified subject and object = "any edge" use-case
    }
  } else if (isFromSelf) {
    if (isToSelf) {
      // Deterministic get/set/remove content == IDX
    } else {
      // Deterministic get/set/remove linkTo, edge content based on schema, ex: account link
    }
  } else if (isToSelf) {
    // Deterministic link with specified subject but object is stream content - what use case?
  } else {
    // Deterministic link with specified subject and object - what use case?
  }
}

type PosterModel = {
  v: 1
  protocols: {
    schemas: {
      BasicProfile: {}
      Post: {}
      EmptySchema: {}
    }
    definitions: {
      basicProfile: {
        // read/write: deterministic/IDX
        schema: '<BasicProfile>'
      }
      post: {
        // read: connection with optional streamID filter (false | string)
        // write: add to connection, optional linkTo param
        schema: '<Post>'
        config: {
          multiple: true
          to: { type: 'streamid'; optional: true } // Can be a reply or original post
        }
      }
      follow: {
        // read: based on DID type, connection from + connection to
        // write: add/remove follow link
        schema: '<EmptySchema>'
        config: {
          multiple: false
          to: { type: 'did' }
        }
      }
    }
  }
}
