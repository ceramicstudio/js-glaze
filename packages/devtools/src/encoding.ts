import type { EncodedDagJWS, EncodedDagJWSResult, ManagedEntry, ManagedModel } from '@glazed/types'
import CID from 'cids'
import type { DagJWS, DagJWSResult } from 'dids'
import { fromString, toString } from 'uint8arrays'

import { applyMap } from './utils'

export function decodeDagJWS({ payload, signatures, link }: EncodedDagJWS): DagJWS {
  return { payload, signatures, link: link ? new CID(link) : undefined }
}

export function encodeDagJWS({ payload, signatures, link }: DagJWS): EncodedDagJWS {
  return { payload, signatures, link: link?.toString() }
}

export function decodeDagJWSResult({ jws, linkedBlock }: EncodedDagJWSResult): DagJWSResult {
  return { jws: decodeDagJWS(jws), linkedBlock: fromString(linkedBlock, 'base64pad') }
}

export function encodeDagJWSResult({ jws, linkedBlock }: DagJWSResult): EncodedDagJWSResult {
  return { jws: encodeDagJWS(jws), linkedBlock: toString(linkedBlock, 'base64pad') }
}

/** @internal */
export function decodeSignedMap<K extends string>(
  data: Record<K, Array<EncodedDagJWSResult>>
): Record<K, Array<DagJWSResult>> {
  return applyMap(data, (records) => records.map(decodeDagJWSResult))
}

/** @internal */
export function encodeSignedMap<K extends string>(
  data: Record<K, Array<DagJWSResult>>
): Record<K, Array<EncodedDagJWSResult>> {
  return applyMap(data, (records) => records.map(encodeDagJWSResult))
}

/** @internal */
export function decodeEntryCommits(
  entry: ManagedEntry<EncodedDagJWSResult>
): ManagedEntry<DagJWSResult> {
  return { ...entry, commits: entry.commits.map(decodeDagJWSResult) }
}

export function decodeModel(model: ManagedModel<EncodedDagJWSResult>): ManagedModel<DagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(decodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, decodeEntryCommits),
    tiles: applyMap(model.tiles, decodeEntryCommits),
  }
}

/** @internal */
export function encodeEntryCommits(
  entry: ManagedEntry<DagJWSResult>
): ManagedEntry<EncodedDagJWSResult> {
  return { ...entry, commits: entry.commits.map(encodeDagJWSResult) }
}

export function encodeModel(model: ManagedModel<DagJWSResult>): ManagedModel<EncodedDagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(encodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, encodeEntryCommits),
    tiles: applyMap(model.tiles, encodeEntryCommits),
  }
}
