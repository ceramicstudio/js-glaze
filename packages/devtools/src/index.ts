/**
 * Development tools library.
 *
 * @module devtools
 */

export * from './composite.js'
export * from './formats/json.js'
export * from './formats/runtime.js'
export * from './utils.js'
