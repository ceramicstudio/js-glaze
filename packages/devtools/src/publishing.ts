import type {
  CeramicApi,
  CeramicCommit,
  GenesisCommit,
  StreamMetadata,
} from '@ceramicnetwork/common'
import { TileDocument } from '@ceramicnetwork/stream-tile'

import { promiseMap } from './utils'

const PUBLISH_OPTS = { anchor: false }

/** @internal */
export async function createModelDoc<T = Record<string, any>>(
  ceramic: CeramicApi,
  content: T,
  metadata: Partial<StreamMetadata> = {}
): Promise<TileDocument<T>> {
  const doc = await TileDocument.create<T>(ceramic, content, metadata, PUBLISH_OPTS)
  await ceramic.pin.add(doc.id)
  return doc
}

/** @internal */
export async function publishCommits(
  ceramic: CeramicApi,
  [genesis, ...updates]: Array<CeramicCommit>
): Promise<TileDocument<Record<string, any>>> {
  const doc = await TileDocument.createFromGenesis<TileDocument<Record<string, any>>>(
    ceramic,
    genesis as GenesisCommit,
    PUBLISH_OPTS
  )
  await ceramic.pin.add(doc.id)
  for (const commit of updates) {
    await ceramic.applyCommit(doc.id, commit, PUBLISH_OPTS)
  }
  return doc
}

/** @internal */
export async function publishSignedMap<T extends string = string>(
  ceramic: CeramicApi,
  signed: Record<T, Array<CeramicCommit>>
): Promise<Record<T, TileDocument>> {
  return await promiseMap(signed, async (commits) => await publishCommits(ceramic, commits))
}
