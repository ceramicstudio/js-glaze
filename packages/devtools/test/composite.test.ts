import type { ModelDefinition } from '@glazed/types'

import { Composite, type CompositeParams } from '../src'

describe('composite', () => {
  describe('Composite instance', () => {
    describe('constructor()', () => {
      test('throws if the version is not compatible', () => {
        expect(
          () => new Composite({ commits: {}, definition: { version: '2.0', models: {} } })
        ).toThrow('Unsupported Composite version 2.0, expected version 1.0')
      })

      test('throws if commits do not match the definition models', () => {
        expect(() => {
          new Composite({
            commits: { fooID: [] },
            definition: {
              version: '1.1',
              models: {
                fooID: {} as unknown as ModelDefinition,
                barID: {} as unknown as ModelDefinition,
              },
            },
          })
        }).toThrow('Missing commits for model barID')
      })
    })

    test('hash() getter returns a stable hash', () => {
      const source = new Composite({
        commits: { fooID: [], barID: [] },
        definition: {
          version: '1.0',
          models: {
            fooID: { name: 'Foo', accountRelation: 'link', schema: {} },
            barID: { name: 'Bar', accountRelation: 'link', schema: {} },
          },
          aliases: { fooID: 'Test', barID: 'Other' },
        },
      })
      const hash = source.hash
      const clone = source.clone()
      expect(clone.hash).toBe(hash)
      expect(source.hash).toBe(hash)
    })

    describe('equals()', () => {
      const params: CompositeParams = {
        commits: { fooID: [], barID: [] },
        definition: {
          version: '1.0',
          models: {
            fooID: { name: 'Foo', accountRelation: 'link', schema: {} },
            barID: { name: 'Bar', accountRelation: 'link', schema: {} },
          },
          aliases: { fooID: 'Test', barID: 'Other' },
        },
      }
      const source = new Composite(params)

      test('with CompositeParams', () => {
        expect(source.equals(params)).toBe(true)
        expect(source.equals({ commits: {}, definition: { version: '1.0', models: {} } })).toBe(
          false
        )
      })

      test('with Composite instance', () => {
        expect(source.equals(source.clone())).toBe(true)
        expect(source.equals(source.setAliases({ fooID: 'Test' }))).toBe(true)
        expect(source.equals(source.setAliases({ bazID: 'Baz' }))).toBe(false)
      })
    })

    test('toJSON() returns a JSON-encoded composite', () => {
      const composite = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
      const encoded = composite.toJSON()
      expect(encoded).toMatchSnapshot()
    })

    test('toParams() returns copies of the Composite commits and definition', () => {
      const inputParams = { commits: {}, definition: { version: '1.0', models: {} } }
      const composite = new Composite(inputParams)
      expect(composite.toParams()).toMatchObject(inputParams)
    })

    test('toRuntime() returns the runtime definition for the Composite', () => {
      const composite = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
      const runtime = composite.toRuntime()
      expect(runtime).toMatchSnapshot()
    })

    test('clone() returns a cloned Composite instance', () => {
      const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
      const sourceParams = source.toParams()
      const clone = source.clone()
      expect(clone.toParams()).toEqual(sourceParams)
    })

    describe('copy() creates a copy of the composite with only selected models', () => {
      const source = new Composite({
        commits: { fooID: [], barID: [] },
        definition: {
          version: '1.0',
          models: {
            fooID: { name: 'Foo', accountRelation: 'link', schema: {} },
            barID: { name: 'Bar', accountRelation: 'link', schema: {} },
          },
          aliases: { fooID: 'Test', barID: 'Other' },
        },
      })

      test('throws an error if no model is set', () => {
        expect(() => {
          source.copy([])
        }).toThrow('Missing models to copy')
      })

      test('throws an error if a model is not found', () => {
        expect(() => {
          source.copy(['unknown'])
        }).toThrow('Model not found: unknown')
      })

      test('identifies models by ID', () => {
        const copy = source.copy(['fooID'])
        expect(copy.toParams()).toEqual({
          commits: { fooID: [] },
          definition: {
            version: '1.0',
            models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
            aliases: { fooID: 'Test' },
            commonEmbeds: [],
            views: { account: {}, models: {}, root: {} },
          },
        })
      })

      test('identifies models by name', () => {
        const copy = source.copy(['Foo'])
        expect(copy.toParams()).toEqual({
          commits: { fooID: [] },
          definition: {
            version: '1.0',
            models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
            aliases: { fooID: 'Test' },
            commonEmbeds: [],
            views: { account: {}, models: {}, root: {} },
          },
        })
      })

      test('identifies models by alias', () => {
        const copy = source.copy(['Test'])
        expect(copy.toParams()).toEqual({
          commits: { fooID: [] },
          definition: {
            version: '1.0',
            models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
            aliases: { fooID: 'Test' },
            commonEmbeds: [],
            views: { account: {}, models: {}, root: {} },
          },
        })
      })
    })

    describe('setAliases()', () => {
      test('merging with existing', () => {
        const source = new Composite({
          commits: {},
          definition: { version: '1.0', models: {}, aliases: { fooID: 'foo', barID: 'bar' } },
        })
        const copy = source.setAliases({ barID: 'test', bazID: 'baz' })
        expect(source.toParams().definition.aliases).toEqual({ fooID: 'foo', barID: 'bar' })
        expect(copy.toParams().definition.aliases).toEqual({
          fooID: 'foo',
          barID: 'test',
          bazID: 'baz',
        })
      })

      test('replacing existing', () => {
        const source = new Composite({
          commits: {},
          definition: { version: '1.0', models: {}, aliases: { fooID: 'foo', barID: 'bar' } },
        })
        const copy = source.setAliases({ barID: 'test', bazID: 'baz' }, true)
        expect(source.toParams().definition.aliases).toEqual({ fooID: 'foo', barID: 'bar' })
        expect(copy.toParams().definition.aliases).toEqual({ barID: 'test', bazID: 'baz' })
      })
    })

    describe('setCommonEmbeds()', () => {
      test('merging with existing', () => {
        const source = new Composite({
          commits: {},
          definition: { version: '1.0', models: {}, commonEmbeds: ['Foo', 'Bar'] },
        })
        const copy = source.setCommonEmbeds(['Bar', 'Baz'])
        expect(source.toParams().definition.commonEmbeds).toEqual(['Foo', 'Bar'])
        expect(copy.toParams().definition.commonEmbeds).toEqual(['Foo', 'Bar', 'Baz'])
      })

      test('replacing existing', () => {
        const source = new Composite({
          commits: {},
          definition: { version: '1.0', models: {}, commonEmbeds: ['Foo', 'Bar'] },
        })
        const copy = source.setCommonEmbeds(['Bar', 'Baz'], true)
        expect(source.toParams().definition.commonEmbeds).toEqual(['Foo', 'Bar'])
        expect(copy.toParams().definition.commonEmbeds).toEqual(['Bar', 'Baz'])
      })
    })

    describe('setViews()', () => {
      test('merging with existing', () => {
        const source = new Composite({
          commits: {},
          definition: {
            version: '1.0',
            models: {},
            views: {
              account: { foo: {} },
              models: {
                model: {
                  foo: {
                    type: 'ReferencedFrom',
                    model: 'foo',
                    property: 'foo',
                    collection: false,
                  },
                  bar: {
                    type: 'ReferencedFrom',
                    model: 'bar',
                    property: 'bar',
                    collection: false,
                  },
                },
              },
              root: { foo: {} },
            },
          },
        })
        const copy = source.setViews({
          account: {},
          models: {
            model: {
              bar: {
                type: 'ReferencedFrom',
                model: 'test',
                property: 'test',
                collection: false,
              },
            },
          },
          root: { foo: 'test', bar: 'test' },
        })
        expect(source.toParams().definition.views).toEqual({
          account: { foo: {} },
          models: {
            model: {
              foo: {
                type: 'ReferencedFrom',
                model: 'foo',
                property: 'foo',
                collection: false,
              },
              bar: {
                type: 'ReferencedFrom',
                model: 'bar',
                property: 'bar',
                collection: false,
              },
            },
          },
          root: { foo: {} },
        })
        expect(copy.toParams().definition.views).toEqual({
          account: { foo: {} },
          models: {
            model: {
              foo: {
                type: 'ReferencedFrom',
                model: 'foo',
                property: 'foo',
                collection: false,
              },
              bar: {
                type: 'ReferencedFrom',
                model: 'test',
                property: 'test',
                collection: false,
              },
            },
          },
          root: { foo: 'test', bar: 'test' },
        })
      })
    })

    describe('merge() merges composites into the instance', () => {
      test('supports Composite instances and params inputs', () => {
        const source = new Composite({
          commits: { fooID: [] },
          definition: {
            version: '1.0',
            models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
          },
        })
        const otherInstance = new Composite({
          commits: { barID: [] },
          definition: {
            version: '1.0',
            models: { barID: { name: 'Bar', accountRelation: 'link', schema: {} } },
          },
        })
        const otherParams: CompositeParams = {
          commits: { bazID: [] },
          definition: {
            version: '1.0',
            models: { bazID: { name: 'Baz', accountRelation: 'link', schema: {} } },
            aliases: { bazID: 'Test' },
          },
        }
        const composite = source.merge([otherInstance, otherParams])
        expect(composite.toParams()).toEqual({
          commits: { fooID: [], barID: [], bazID: [] },
          definition: {
            version: '1.0',
            models: {
              fooID: { name: 'Foo', accountRelation: 'link', schema: {} },
              barID: { name: 'Bar', accountRelation: 'link', schema: {} },
              bazID: { name: 'Baz', accountRelation: 'link', schema: {} },
            },
            aliases: { bazID: 'Test' },
            commonEmbeds: [],
            views: {
              account: {},
              models: {},
              root: {},
            },
          },
        })
      })

      test('throws if any input uses an unsupported version', () => {
        const composite = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const other = { commits: {}, definition: { version: '2.0', models: {} } }
        expect(() => composite.merge(other)).toThrow(
          'Unsupported Composite version 2.0, expected version 1.0'
        )
      })

      test('throws if any input does not have matching commits', () => {
        const composite = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const other: CompositeParams = {
          commits: {},
          definition: {
            version: '1.5',
            models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
          },
        }
        expect(() => composite.merge(other)).toThrow('Missing commits for model fooID')
      })

      test('with aliases', () => {
        const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const composite = source.merge(
          [
            {
              commits: {},
              definition: { version: '1.0', models: {}, aliases: { fooID: 'Foo', barID: 'Bar' } },
            },
            {
              commits: {},
              definition: { version: '1.0', models: {}, aliases: { barID: 'Test', bazID: 'Baz' } },
            },
          ],
          { aliases: { bazID: 'Test' } }
        )
        expect(composite.toParams().definition.aliases).toEqual({
          fooID: 'Foo',
          barID: 'Test',
          bazID: 'Test',
        })
      })

      test('with no common embeds (default)', () => {
        const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const composite = source.merge([
          {
            commits: {},
            definition: { version: '1.0', models: {}, commonEmbeds: ['Foo', 'Bar'] },
          },
          { commits: {}, definition: { version: '1.0', models: {} } },
        ])
        expect(composite.toParams().definition.commonEmbeds).toHaveLength(0)
      })

      test('with all common embeds', () => {
        const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const composite = source.merge(
          [
            {
              commits: {},
              definition: { version: '1.0', models: {}, commonEmbeds: ['Foo', 'Bar'] },
            },
            {
              commits: {},
              definition: { version: '1.0', models: {}, commonEmbeds: ['Bar', 'Baz'] },
            },
          ],
          { commonEmbeds: 'all' }
        )
        expect(composite.toParams().definition.commonEmbeds).toEqual(['Foo', 'Bar', 'Baz'])
      })

      test('with specific common embeds', () => {
        const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const composite = source.merge(
          [
            {
              commits: {},
              definition: { version: '1.0', models: {}, commonEmbeds: ['Foo', 'Bar'] },
            },
            {
              commits: {},
              definition: { version: '1.0', models: {}, commonEmbeds: ['Bar', 'Baz'] },
            },
          ],
          { commonEmbeds: ['Some', 'Other'] }
        )
        expect(composite.toParams().definition.commonEmbeds).toEqual(['Some', 'Other'])
      })

      test('with views', () => {
        const source = new Composite({ commits: {}, definition: { version: '1.0', models: {} } })
        const composite = source.merge(
          [
            {
              commits: {},
              definition: {
                version: '1.0',
                models: {},
                views: { account: { foo: {} }, models: {}, root: {} },
              },
            },
            {
              commits: {},
              definition: {
                version: '1.0',
                models: {},
                views: { account: { bar: {} }, models: {}, root: { foo: {} } },
              },
            },
          ],
          { views: { account: {}, models: {}, root: { foo: 'test' } } }
        )
        expect(composite.toParams().definition.views).toEqual({
          account: { foo: {}, bar: {} },
          models: {},
          root: { foo: 'test' },
        })
      })
    })
  })

  test('Composite.compose() merges composites into a new instance', () => {
    const first: CompositeParams = {
      commits: { fooID: [] },
      definition: {
        version: '1.0',
        commonEmbeds: ['First'],
        models: { fooID: { name: 'Foo', accountRelation: 'link', schema: {} } },
      },
    }
    const second: CompositeParams = {
      commits: { barID: [] },
      definition: {
        version: '1.0',
        commonEmbeds: ['First', 'Second'],
        models: { barID: { name: 'Bar', accountRelation: 'link', schema: {} } },
      },
    }
    const third: CompositeParams = {
      commits: { bazID: [] },
      definition: {
        version: '1.0',
        models: { bazID: { name: 'Baz', accountRelation: 'link', schema: {} } },
        aliases: { bazID: 'Test' },
      },
    }
    const composite = Composite.compose([new Composite(first), second, third], {
      commonEmbeds: 'all',
    })
    const params = composite.toParams()
    expect(params).toEqual({
      commits: { fooID: [], barID: [], bazID: [] },
      definition: {
        version: '1.0',
        models: {
          fooID: { name: 'Foo', accountRelation: 'link', schema: {} },
          barID: { name: 'Bar', accountRelation: 'link', schema: {} },
          bazID: { name: 'Baz', accountRelation: 'link', schema: {} },
        },
        aliases: { bazID: 'Test' },
        commonEmbeds: ['First', 'Second'],
        views: {
          account: {},
          models: {},
          root: {},
        },
      },
    })
    // Source composites should not be altered
    expect(params).not.toEqual(first)
    expect(params).not.toEqual(second)
    expect(params).not.toEqual(third)
  })
})
