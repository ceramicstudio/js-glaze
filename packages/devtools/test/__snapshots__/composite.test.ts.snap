// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composite Composite instance toJSON() returns a JSON-encoded composite 1`] = `
Object {
  "aliases": Object {},
  "models": Object {},
  "version": "1.0",
  "views": Object {
    "account": Object {},
    "models": Object {},
    "root": Object {},
  },
}
`;

exports[`composite Composite instance toRuntime() returns the runtime definition for the Composite 1`] = `
Object {
  "accountStore": Object {},
  "models": Object {},
  "objects": Object {},
}
`;
